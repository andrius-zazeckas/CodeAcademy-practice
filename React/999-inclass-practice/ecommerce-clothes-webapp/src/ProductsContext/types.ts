export type TProduct = {
  id: number;
  title: string | null;
  price: number | null;
  description: string | null;
  category: string | null;
  image: string;
  rating: {
    rate: number | null;
    count: number | null;
  };
};

export type TCartProduct = TProduct & {
  amount: number;
};

export type TProductsContext = {
  fetchedProducts: Readonly<TProduct[]>;
  cartProducts: Readonly<TCartProduct[]>;
  //   dispatch: () => void;
  dispatch(): void;
};
