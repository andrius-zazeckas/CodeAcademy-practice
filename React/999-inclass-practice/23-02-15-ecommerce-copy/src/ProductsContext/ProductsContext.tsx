import { createContext } from "react";
import { TProductsContext } from "./types";

export const ProductsContext = createContext<TProductsContext>({
  products: [],
  setProducts: () => {},
});
