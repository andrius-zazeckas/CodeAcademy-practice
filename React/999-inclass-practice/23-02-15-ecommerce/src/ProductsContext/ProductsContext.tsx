import { createContext } from "react";
import type { TProductsContext } from "./types";

export const INITIAL_VALUE: TProductsContext = {
  products: [],
  setProducts: () => {},
  cartProducts: [],
  setCartProducts: () => {},
};

// const INITIAL_VALUE = {
//   products: [],
//   setProducts: () => {},
//   cartProducts: [],
//   setCartProducts: () => {},
// } satisfies TProductsContext;

// const INITIAL_VALUE = {
//   products: [],
//   setProducts: () => {},
//   cartProducts: [],
//   setCartProducts: () => {},
// } as TProductsContext;

export const ProductsContext = createContext(INITIAL_VALUE);
