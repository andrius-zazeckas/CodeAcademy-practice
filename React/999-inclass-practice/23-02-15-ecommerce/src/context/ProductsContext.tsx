import { createContext } from "react";
import { TProducts } from "../types/Tproducts";

export const ProductsContext = createContext<TProducts[]>([]);
