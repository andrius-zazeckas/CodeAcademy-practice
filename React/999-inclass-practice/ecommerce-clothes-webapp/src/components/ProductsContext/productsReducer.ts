import type { TProductsAction, TProductsState } from "./types";

export const productsReducer = (
  state: TProductsState,
  action: TProductsAction
) => {
  switch (action.type) {
    case "addProduct":
      console.log("add product");
      break;

    case "removeProduct":
      console.log("remove product");
      break;

    case "setProducts":
      console.log("set products");
      break;

    default:
      console.log("no case matched");
  }
  return state;
};
