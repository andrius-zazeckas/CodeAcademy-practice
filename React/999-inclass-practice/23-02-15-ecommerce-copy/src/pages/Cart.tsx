import { useContext } from "react";
import { CartProductsContext } from "../ProductsContext/CartProductsContext";
import { ProductContainer } from "../styles/ProductContainer";
import { ProductsContainer } from "../styles/ProductsContainer";
import { TransparentButton } from "../styles/TransparentButton";

export const Cart = () => {
  const { cartProducts, setCartProducts } = useContext(CartProductsContext);

  const handleIncrement = (productIndex: number) => {
    const modifiedProducts = [...cartProducts];
    const product = modifiedProducts[productIndex];

    product.amount = ++product.amount;

    setCartProducts(modifiedProducts);
  };

  const handleDecrement = (productIndex: number) => {
    const modifiedProducts = [...cartProducts];
    const product = modifiedProducts[productIndex];

    product.amount -= 1;

    if (!product.amount) {
      cartProducts.splice(productIndex, 1);

      return setCartProducts([...cartProducts]);
    }

    setCartProducts(modifiedProducts);
  };

  return (
    <ProductsContainer>
      {cartProducts.map((product: any, i: number) => (
        <ProductContainer key={product.id}>
          <p>{product.title}</p>
          <p>Product amount: {product.amount}</p>
          {product.price ? (
            <>
              <p>Product price: {product.price}</p>
              <p>Sum: {(product.price * product.amount).toFixed(2)}</p>
            </>
          ) : null}
          <TransparentButton onClick={() => handleIncrement(i)}>
            +
          </TransparentButton>
          <TransparentButton onClick={() => handleDecrement(i)}>
            -
          </TransparentButton>
        </ProductContainer>
      ))}
    </ProductsContainer>
  );
};
