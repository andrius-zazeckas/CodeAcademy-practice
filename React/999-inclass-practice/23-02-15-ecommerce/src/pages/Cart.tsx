import { useContext } from "react";
import { CartProductsContext } from "../ProductsContext/CartProductsContext";
import { ProductContainer } from "../styles/ProductContainer";
import { ProductsContainer } from "../styles/ProductsContainer";
import { TransparentButton } from "../styles/TransparentButton";

export const Cart = () => {
  const { cartProducts, setCartProducts } = useContext(CartProductsContext);

  const handlePlus = (productIndex: number) => {
    const product = cartProducts[productIndex];
    // const isProductInCart = cartProducts.some(
    //   (cartProduct) => cartProduct.id === product.id
    // );

    product.amount = ++product.amount;

    // if (isProductInCart) {
    //   return setCartProducts((prevCartProducts) => {
    //     const newCartProducts = prevCartProducts.map((cartProduct) => ({
    //       ...cartProduct,
    //       amount:
    //         cartProduct.id === product.id
    //           ? cartProduct.amount + 1
    //           : cartProduct.amount,
    //     }));
    //     return newCartProducts;
    //   });
    // }

    setCartProducts([...cartProducts]);
  };

  const handleMinus = (productIndex: number) => {
    const product = cartProducts[productIndex];

    product.amount -= 1;

    if (product.amount === 0) {
      cartProducts.splice(productIndex, 1);

      return setCartProducts([...cartProducts]);
    }

    setCartProducts([...cartProducts]);
  };

  return (
    <ProductsContainer>
      {cartProducts.map((product: any, i: number) => (
        <ProductContainer key={product.id}>
          <p>{product.title}</p>
          <p>Product amount: {product.amount}</p>
          <p>Product price: {product.price}</p>
          <p>Sum: {(product.price * product.amount).toFixed(2)}</p>
          <TransparentButton onClick={() => handlePlus(i)}>+</TransparentButton>
          <TransparentButton onClick={() => handleMinus(i)}>
            -
          </TransparentButton>
        </ProductContainer>
      ))}
    </ProductsContainer>
  );
};
