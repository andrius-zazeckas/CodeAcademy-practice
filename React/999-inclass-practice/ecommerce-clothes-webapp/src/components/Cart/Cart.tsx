import { useContext } from "react";
import { ProductsContext } from "../ProductsContext";

export const Cart = () => {
  const { dispatch } = useContext(ProductsContext);

  return <></>;
};
