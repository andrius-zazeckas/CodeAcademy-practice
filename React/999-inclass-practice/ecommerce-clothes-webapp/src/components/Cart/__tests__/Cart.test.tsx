import { render, screen } from "@testing-library/react";
import { Cart } from "../Cart";
import renderer from "react-test-renderer";
import { ProductsContext } from "../../ProductsContext";

describe("Cart", () => {
  const cartProducts = [
    {
      id: 1,
      title: "Test",
      price: 2.99,
      description: "For testing purposes",
      category: "Testing category",
      image: "testImage.jpg",
      rating: {
        rate: 4,
        count: 2,
      },
      amount: 2,
    },
  ];

  it("should render Cart", () => {
    render(<Cart />);
    expect(screen.getByLabelText("cart")).toBeVisible();

    const tree = renderer.create(<Cart />);

    expect(tree).toMatchSnapshot();
  });

  it("should show product amount in cart", () => {
    render(
      <ProductsContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    expect(screen.getByLabelText("product amount").textContent).toBe(
      `${cartProducts[0].amount}`
    );

    const tree = renderer.create(
      <ProductsContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    expect(tree).toMatchSnapshot();
  });

  it("should calculate cart total price", () => {
    render(
      <ProductsContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    const currencyFormat = new Intl.NumberFormat(navigator.language, {
      style: "currency",
      currency: "EUR",
    });

    const totalPrice = cartProducts[0].amount * cartProducts[0].price;

    expect(screen.getByLabelText("total price").textContent).toBe(
      `${currencyFormat.format(totalPrice)}`
    );

    const tree = renderer.create(
      <ProductsContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    expect(tree).toMatchSnapshot();
  });
});
