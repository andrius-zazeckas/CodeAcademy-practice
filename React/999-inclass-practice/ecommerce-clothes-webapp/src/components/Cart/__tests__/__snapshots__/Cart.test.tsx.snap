// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should calculate cart total price 1`] = `
<div
  aria-label="cart"
  className="MuiBox-root css-0"
>
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        aria-label="cart product"
        className="MuiGrid-root MuiGrid-container MuiGrid-item css-hayvbu-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Test
          </p>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-82ndot-MuiGrid-root"
        >
          <button
            aria-label="add product"
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1klp1e9-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            +
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <p
            aria-label="product amount"
            className="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
          >
            2
          </p>
          <button
            aria-label="remove product"
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1orhlu6-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            -
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-hlnzui-MuiTypography-root"
          >
            €2.99
          </p>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root css-1ez87es"
    >
      <h4
        className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      >
        Total price: 
      </h4>
      <h4
        aria-label="total price"
        className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      >
        €5.98
      </h4>
    </div>
  </div>
</div>
`;

exports[`Cart should render Cart 1`] = `
<div
  aria-label="cart"
  className="MuiBox-root css-0"
>
  <h3
    className="MuiTypography-root MuiTypography-body1 css-1ykj1nk-MuiTypography-root"
  >
    No products in cart
  </h3>
</div>
`;

exports[`Cart should show product amount in cart 1`] = `
<div
  aria-label="cart"
  className="MuiBox-root css-0"
>
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        aria-label="cart product"
        className="MuiGrid-root MuiGrid-container MuiGrid-item css-hayvbu-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Test
          </p>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-82ndot-MuiGrid-root"
        >
          <button
            aria-label="add product"
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1klp1e9-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            +
          </button>
          <p
            aria-label="product amount"
            className="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
          >
            2
          </p>
          <button
            aria-label="remove product"
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1orhlu6-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            -
          </button>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-hlnzui-MuiTypography-root"
          >
            €2.99
          </p>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root css-1ez87es"
    >
      <h4
        className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      >
        Total price: 
      </h4>
      <h4
        aria-label="total price"
        className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      >
        €5.98
      </h4>
    </div>
  </div>
</div>
`;
